crates/wasmtime/src/runtime/ref/gc_ref.rs

```diff
@@ -0,0 +1,107 @@
+use crate::AsContextMut;
+use std::any::Any;
+use std::ffi::c_void;
+use wasmtime_runtime::VMExternRef;
+
+/// Represents an opaque reference to any data within WebAssembly.
+#[derive(Clone, Debug)]
+#[repr(transparent)]
+pub struct ExternRef {
+    pub(crate) inner: VMExternRef,
+}
+
+impl ExternRef {
+    /// Creates a new instance of `ExternRef` wrapping the given value.
+    pub fn new<T>(value: T) -> ExternRef
+    where
+        T: 'static + Any + Send + Sync,
+    {
+        let inner = VMExternRef::new(value);
+        ExternRef { inner }
+    }
+
+    /// Get the underlying data for this `ExternRef`.
+    pub fn data(&self) -> &dyn Any {
+        &*self.inner
```

Comment:
I actually think we want to keep whole implementations of `ExternRef` separate for GC vs not GC. Looking into the future, there are going to be a bunch more types that are going to be doing this same split, likely whole submodules of things, and I don't think it will be very nice to have all these `cfg`s for the whole submodule tree. I think it will be better to have (a) the GC-enabled submodule tree and (b) a single GC-disabled module that defines all the same things from (a) but with uninhabited types.

Reply:
They do? The only ones left throughout the code base should be things that call constructors.

Reply:
(It just turns out we call constructors frequently, unfortunately)